src/Analyzers/CSharp/Analyzers/RemoveUnusedMembers/CSharpRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -10,7 +13,20 @@ namespace Microsoft.CodeAnalysis.CSharp.RemoveUnusedMembers
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     internal class CSharpRemoveUnusedMembersDiagnosticAnalyzer
-        : AbstractRemoveUnusedMembersDiagnosticAnalyzer<DocumentationCommentTriviaSyntax, IdentifierNameSyntax>
+        : AbstractRemoveUnusedMembersDiagnosticAnalyzer<
+            DocumentationCommentTriviaSyntax,
+            IdentifierNameSyntax,
+            TypeDeclarationSyntax,
+            MemberDeclarationSyntax>
     {
+        protected override IEnumerable<TypeDeclarationSyntax> GetTypeDeclarations(INamedTypeSymbol namedType, CancellationToken cancellationToken)
+        {
+            return namedType.DeclaringSyntaxReferences
+                .Select(r => r.GetSyntax(cancellationToken))
+                .OfType<TypeDeclarationSyntax>();
+        }
+
+        protected override SyntaxList<MemberDeclarationSyntax> GetMembers(TypeDeclarationSyntax typeDeclaration)
+            => typeDeclaration.Members;
```

Comment:
all the core walking logic is in the base type.  the subtypes just need to indicate how to get the type-decl nodes for a particular named type, and then how to get the members from those typedecls.